### PR TITLE
v3.0.0 Deploy CMS inside the gitroot

### PR DESCRIPTION
**Pull request by CZYK**
# Changed behaviour:
### Deploying inside the git-root instead of towards deploy-location.
The deploy happens now inside the git-root. This means that we should ignore the artefacts of the deploy. For instance the compiles js/css in the `dist` folder, configuration, cache, error logs etc.

## Configuration:
Previously the configuration was copied like this:
- `Deploy/Config/Live => ./`
- `Deploy/Config/<ENV> => ./` _happened only when the env was not set to Production or Live_

The `Live` config was always copied and in many cases used to copy configuration files which applies to all environments (except when overwritten). It turned out to be more a mix of Base config and environment specific config. That is why the configuration step has been changed to copy the files like this:
- `Deploy/Config/Base => ./`
- `Deploy/Config/<ENV> => ./`

The env Production is not renamed to Live anymore so the folder name should be Production, this brings it more in line with the actual env names we use.

The quickest way to port your product's config is to rename the `Deploy\Config\Live` folder to `Deploy\Config\Base`. After this step you can extract the env specific config from  `Deploy\Config\Base` and move them to  `Deploy\Config\Production`. This is not mandatory but nice to have the config files actually where they belong.

## Watcher
The cms watcher exposes itself as a separate task and is not run automatically anymore on `gulp.stop`. This means that the user decided which task to execute. I propose to use a npm script `npm start` to execute the required tasks in the desired order. For instance:
```json
  "scripts": {
    "start": "vagrant.exe up && gulp deploy.cms && npm run open && gulp watch.cms",
  },
```

## Dependency management
The choice has been made to fix sensitive dependencies (like sass, js) on exact version instead of 'compatible with'. By doing this we make sure that nothing changes in our critical deployment process. Minor and patch releases should not change behaviour but we cannot control how strict the owners of third party packages follow up on this.

See https://github.com/studyportals/product-deploy/issues/84 for an example. For us this resulted in breaking deploys for a windows of 2 hours. It's just a small example which shows how a patch release, which appears to be non-breaking, in some situation does break. Proper test-cases minimise the chance of releasing a breaking change but it is no guarantee the test-cases will cover all use-cases.

# Solved issues:
- resolves studyportals/product-deploy/issues/63
- resolves studyportals/product-deploy/issues/58 (kinda)
- resolves studyportals/product-deploy/issues/82
- resolves studyportals/product-deploy/issues/54
- resolves studyportals/product-deploy#86
- fixes studyportals/product-deploy/issues/62
- fixes studyportals/product-deploy/issues/84

### CODE REVIEW

File: lib/private/write-revision.json.js [hunk 1]
```diff
@@ -3,6 +3,19 @@
 const path = require('path');
 const fs = require('fs-extra');
 
+const getProduct = (src) => {
+
+	const pjson = path.resolve(src, "package.json");
+
+	if(fs.existsSync(pjson)){
+
+		return require(pjson).repository.url.split('/').slice(-1)[0].replace(".git", "");
```
Comment (NikolaNikushev) [inline lib/private/write-revision.json.js]: You could use the pjson.name.split('/')[1]
and if necessary to capitalize the first letter ;)

Comment (iliqportals) [reply lib/private/write-revision.json.js]: repository.url is less prone to change

File: lib/simple-page-tester.js [hunk 2]
```diff
@@ -43,16 +43,23 @@ class SimplePageTester {
...
 						.get(path)
 						.set('Cache-Control', 'no-cache')
 						.end((err, res) =>{
 
-							chai.expect(err).to.not.exist;
-							res.should.have.status(200);
-							done();
+							if(res.status === 503){
+
+								this.skip();
```
Comment (iliqportals) [inline lib/simple-page-tester.js]: I would recommend retries.
Skip a test if it's 503 would mean we will not know if that page is currently slow, or it is slow by principle.

```
const MAX_PAGE_ATTEMPTS = 3;
const attemptGetPage = (page, attempt) => {

	if(!attempt) attempt = 1
	chai.expect(attempt).to.be.below(MAX_PAGE_ATTEMPTS, `could not get page ${page} after ${MAX_PAGE_ATTEMPTS} attempts`);
	return new Promise(resolve => {

		server
			.get(page)
			.set('Cache-Control', 'no-cache')
			.end((err, res) =>{
				chai.expect(err).to.not.exist;
				if(res.status === 503) return attemptGetPage(page, attempt + 1);
				res.should.have.status(200);
				resolve();
			});
	});
}

it(`${path} should return a 200`, () => {

	return attemptGetPage(path).catch(ex => {

		chai.expect(ex).to.equal(undefined, ex)
	});
});
```

File: lib/deploy.js [hunk 3]
```diff
@@ -6,32 +6,42 @@ const sass = require('./private/sass');
...
  * @type {Deploy}
  */
 class Deploy{
 	/**
 	 * @param {Object} opts
-	 * @param {string} opts.to - Deploy folder
 	 * @param {string} [opts.from=process.cwd()] - Source folder
 	 * @param {Gulp} [opts.gulp=null] - Instance of Gulp
+	 *
+	 * TODO: process.env.PRTL_ENV
```
Comment (iliqportals) [inline lib/deploy.js]: What is intended with these todos?

**Comment by CZYK** [comment]
@iliqportals I have remove the TODO's. Added some in the top so I might want to roll them out as well.

**Pull request merged**